INT-4015: Streaming Remote File Inbound Adapter

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4015
      https://jira.spring.io/browse/INT-3854

Initial commit.

Reworked to emit an input stream and use the file splitter.

Add StreamTransformer.

Add CLOSABLE_RESOURCE header so we can close the session automatically.

Implement INT-3854, FTP, SFTP

garyrussell committed:
INT-4015: Streaming Remote File Inbound Adapter

JIRA: https://jira.spring.io/browse/INT-4015
      https://jira.spring.io/browse/INT-3854

Initial commit.

Reworked to emit an input stream and use the file splitter.

Add StreamTransformer.

Add CLOSABLE_RESOURCE header so we can close the session automatically.

Implement INT-3854, FTP, SFTP

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/remote/AbstractRemoteFileStreamingMessageSource.java:
```diff
@@ -0,0 +1,208 @@
+		if (file != null) {
+			String remotePath = remotePath(file);
+			if (this.deleteRemoteFiles) {
+				this.remoteFileTemplate.remove(remotePath);
```
H-m. How can we `stream` the remote file afterward if we remove it here?

garyrussell replied:
Doh - I'll remove that option - they can use a gateway to rm.

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/splitter/FileSplitter.java:
```diff
@@ -158,7 +160,19 @@ else if (payload instanceof Reader) {
+				if (closeableResource != null) {
+					closeableResource.close();
+				}
+				super.close();
```
Maybe in the `finally`?
It's not еhe `BufferedReader`  fault the we can't close `Session` properly.

garyrussell committed:
(S)FTP Namespace Changes

garyrussell committed:
Docs - also fixes a PDF overflow

garyrussell committed:
Polishing - PR Comments

garyrussell commented:
Pushed

garyrussell committed:
checkstyle fixes

garyrussell commented:
Summary of schema changes:

- `charset` was incorrectly on inbound channel adapters - moved to `file:remoteOutboundAttributeGroup`
- `remote-directory-expression` applies to inbound and outbound - moved from `file:remoteOutboundAttributeGroup` and `inbound-channel-adapter` to `base-adapter-type`
- renamed `base-ftp-adapter-type` to `base-inbound-adapter-type` and created `base-outbound-adapter-type`; moved common inbound attributes to `base-inbound-adapter-type`
- created `inbound-streaming-channel-adapter`
- moved `temporary-file-suffix` from base-adapter-type to a new group - added to `base-outbound-adapter-type` and inbound adapter (not needed by new adapter)

artembilan commented:
Thank you Gary!
Will fight with your PR in the evening.

I like it, as well as the feature we expose with it :smile:

garyrussell committed:
Polishing - Add Namespace for StreamParser

garyrussell commented:
No hurry - added a bit more polish.

garyrussell commented:
>I like it, as well as the feature we expose with it :smile:

Definitely a much simpler solution after your suggestion.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/IntegrationMessageHeaderAccessor.java:
```diff
@@ -76,6 +79,19 @@ public Integer getPriority() {
 
+	/**
+	 * If the payload was created by a {@link Closeable} that needs to remain
+	 * open until the payoad is consumed, the resource will be added to this
```
Typo: pay**l**oad

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/StreamTransformerParser.java:
```diff
@@ -0,0 +1,33 @@
+import org.springframework.integration.transformer.StreamTransformer;
+
+/**
+ * @author Gary Russell
```
JavaDoc to explain the premise of the parser.
Like that one on super class.

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/xml/StreamTransformerParserTests.java:
```diff
@@ -0,0 +1,88 @@
+
+/**
+ * @author Mark Fisher
+ * @author Gary Russell
```
Ha! Mark had time to work here as well :smile:

garyrussell replied:
He gets credit because it's a cloned test 😄

artembilan commented on spring-integration-core/src/main/resources/org/springframework/integration/config/spring-integration-4.3.xsd:
```diff
@@ -2691,6 +2691,34 @@
+		</xsd:annotation>
+		<xsd:complexType>
+			<xsd:complexContent>
+				<xsd:extension base="specialized-transformer-type">
```
I wonder why do we allow something else as sub-elements, but not just `<poller>`:
````xml
	<xsd:complexType name="specialized-transformer-type">
		<xsd:choice minOccurs="1" maxOccurs="unbounded">
			<xsd:any processContents="strict" namespace="##other" minOccurs="0" maxOccurs="unbounded" />
			<xsd:element ref="poller" />
		</xsd:choice>
		<xsd:attribute name="id" type="xsd:string" />
	</xsd:complexType>
````

garyrussell replied:
I based this on object-to-string-transformer - the only difference is the charset is not required here.

So it looks like that's wrong too - will take a look.

artembilan replied:
Oh! Sorry, Gary. I didn't mean that you had made a mistake.
I just proposed to review (and maybe fix) that `specialized-transformer-type`, because it allows (by mistake?) any nested tags for awhile, not only `<poller>`.
That is my concern and I need here one more pair of eyes to be sure that I'm not insane :smile:

garyrussell replied:
I just commented it out and `testAll` ran ok.

artembilan replied:
:+1: no comments.
Just ship it! :smile:

artembilan commented on spring-integration-core/src/main/resources/org/springframework/integration/config/spring-integration-4.3.xsd:
```diff
@@ -2691,6 +2691,34 @@
 		</xsd:complexType>
 	</xsd:element>
 
+	<xsd:element name="stream-transformer">
```
This one must be present in the `<chain>`, too.
See `chain-elements-group`

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/xml/StreamTransformerParserTests-context.xml:
```diff
@@ -0,0 +1,29 @@
+	</stream-transformer>
+
+	<stream-transformer id="withCharset" charset="UTF-8"
+		input-channel="charsetChannel" output-channel="output"/>
```
I think after `<chain>` fix, the test on the matter won't be redundant here.

garyrussell replied:
redundant? This is testing the string option.

artembilan replied:
???
... **won't** be redundant 

I meant we should add the `<chain>` variant with the `<stream-transformer>` to this parser's test.

Note sure how to point easier for adding new changes to the file... 
That's why I did that in the end of this config.

Sorry, I need to be more clear.

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java:
```diff
@@ -57,7 +57,8 @@ protected final BeanMetadataElement parseSource(Element element, ParserContext p
 		IntegrationNamespaceUtils.setValueIfAttributeDefined(synchronizerBuilder, element, "temporary-file-suffix");
-		this.configureFilter(synchronizerBuilder, element, parserContext);
+		FileParserUtils.configureFilter(synchronizerBuilder, element, parserContext,
+				getSimplePatternFileListFilterClassname(), getRegexPatternFileListFilterClassname());
```
If we are refactoring this already, maybe let's switch to `Class<?>`.
My hands itch to fix the `...Classname()` camel-case typo, when I see this code :smile:

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/config/AbstractRemoteFileInboundChannelAdapterParser.java:
```diff
@@ -57,7 +57,8 @@ protected final BeanMetadataElement parseSource(Element element, ParserContext p
 		String remoteFileSeparator = element.getAttribute("remote-file-separator");
 		synchronizerBuilder.addPropertyValue("remoteFileSeparator", remoteFileSeparator);
```
M-m-m. Looks like we can replace these to single line like below:
```
IntegrationNamespaceUtils.setValueIfAttributeDefined(synchronizerBuilder, element, "remote-file-separator");
```

garyrussell replied:
No; this was explicitly added to allow `""`

artembilan replied:
`setValueIfAttributeDefined(BeanDefinitionBuilder builder, Element element, String attributeName,
			String propertyName, boolean emptyStringAllowed)`
then?

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/remote/RemoteFileOperations.java:
```diff
@@ -141,4 +149,11 @@
+	 * needed.
+	 * @return a session.
+	 */
+	Session<F> getSesssion();
```
New methods with `@since 4.3`

artembilan commented on spring-integration-file/src/main/java/org/springframework/integration/file/remote/gateway/AbstractRemoteFileOutboundGateway.java:
```diff
@@ -586,7 +587,8 @@ public File doInSession(Session<F> session) throws IOException {
 				.setHeader(FileHeaders.REMOTE_DIRECTORY, remoteDir)
 				.setHeader(FileHeaders.REMOTE_FILE, remoteFilename)
-				.setHeader(FileHeaders.REMOTE_SESSION, session)
+				.setHeader("file_remoteSession", session) // TODO: remove in 5.0
```
Have started story for similar cases: https://jira.spring.io/browse/INT-4018

artembilan commented on spring-integration-file/src/test/java/org/springframework/integration/file/remote/RemoteFileTestSupport.java:
```diff
@@ -0,0 +1,176 @@
+ * Abstract base class for tests requiring remote file servers, e.g. (S)FTP.
+ *
+ * @author Gary Russell
+ *
```
`@since 4.3`

artembilan commented on spring-integration-file/src/test/java/org/springframework/integration/file/remote/StreamingInboundTests.java:
```diff
@@ -0,0 +1,223 @@
+				return session;
+			}
+			catch (Exception e) {
+				fail("failed to mock session");
```
Why rethrow some `RuntimeException` isn't enough?
Won't this `fail()` hide the real issue, if someone will modify the mocking wrong way?

artembilan commented on spring-integration-ftp/src/test/java/org/springframework/integration/ftp/FtpTestSupport.java:
```diff
@@ -0,0 +1,150 @@
+ * @author Artem Bilan
+ * @author Gary Russell
+ * @author David Turanski
+ */
```
`@since 4.3` ?

artembilan commented on spring-integration-ftp/src/test/java/org/springframework/integration/ftp/inbound/FtpStreamingMessageSourceTests.java:
```diff
@@ -0,0 +1,116 @@
+		}
+
+		@Bean
+		public static SessionFactory<FTPFile> ftpSessionFactory() {
```
Why these two beans are `static`?
Looks like is just overhead.
Although it isn't critical.
Never mind

artembilan commented on src/reference/asciidoc/file.adoc:
```diff
@@ -858,3 +859,8 @@ public MessageHandler fileSplitter() {
 ----
+
+The `FileSplitter` will also split any text-based `InputStream` into lines.
+When used in conjunction with an FTP or SFTP streaming inbound channel adapter, or an FTP or SFTP outbound gateway
```
How about to add cross-links to (S)FTP chapters from here?

artembilan commented on src/reference/asciidoc/sftp.adoc:
```diff
@@ -434,6 +434,41 @@ This will work for any `ResettableFileListFilter`.
+
+[source, xml]
+----
+<int-ftp:inbound-streaming-channel-adapter id="ftpInbound"
```
`sftp` ? :smile:

artembilan commented on src/reference/asciidoc/whats-new.adoc:
```diff
@@ -176,14 +188,26 @@ The SFTP outbound gateway (for `put` and `mput` commands) and the SFTP outbound
+===== Streaming Inbound Channel Adapter
+
+A new inbound channel adapter is provided that returns an `InputStream` for each file.
+See <<ftp-streaming>> or more information.
```
I think `stream` adapters should move to `New Components` section.

artembilan commented:
Somewhere in `ftp/sftp` chapters we have mention of the deprecated `file_remoteSession` header.
I think would be better to redo to a new one `CLOSEABLE_RESOURCE`.

That's all for today.

Excellent PR, it's a pity that it is close for merge :smile:

garyrussell committed:
Polishing - PR Comments

garyrussell commented:
Pushed

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/287d924fc0024dab2833d6c24de8d10d770c8c90
